### DISCUSSION TITLE
How to test a mocked resource in an operation?

### DISCUSSION THREAD

**Discussion by Auric-Manteo** [dagster-io/dagster#18661] [Q&A] [answered]
Hi!
How can I test that a method on a resource was called?
It seems like the resource is not passed directly so I can't simply check if `slack.get_client().chat_postMessage` has been called directly like in this simplified example:
(and yes I would like to pass in the resource via the context :D)

```python
    @op()
    def my_op(context, slack) -> None:
        slack.get_client().chat_postMessage(channel='test',text=f'test')
    
    def test() -> None:
        slack_resource_mock = mock.MagicMock(spec=dict)
        slack_resource_mock.get_client = mock.MagicMock()
        chat_postMessage = mock.MagicMock()
        slack_resource_mock.get_client.return_value = chat_postMessage
    
        context = build_op_context(resources={'slack': slack_resource_mock})
        my_op(context=context)
    
        chat_postMessage.assert_called_once_with(
            channel='test',
            text='test',
        )
```

**Comment by jamiedemaria** [comment, marked as answer]
One potential issue with this code snippet is that the op is getting the `slack` resource as a parameter, but the resource is being passed on the context in the test. You'll want to align on a single method for providing the context to the op. Since you say you want to pass the resource via the context, you'll want to do this 

```python
@op(required_resource_keys={"slack"})
    def my_op(context) -> None:
        context.resources.slack.get_client().chat_postMessage(channel='test',text=f'test')
    
    def test() -> None:
        slack_resource_mock = mock.MagicMock(spec=dict)
        slack_resource_mock.get_client = mock.MagicMock()
        chat_postMessage = mock.MagicMock()
        slack_resource_mock.get_client.return_value = chat_postMessage
    
        context = build_op_context(resources={'slack': slack_resource_mock})
        my_op(context=context)
    
        chat_postMessage.assert_called_once_with(
            channel='test',
            text='test',
        )
```

let me know how that goes

**Reply by Auric-Manteo** [reply]
Your code circumvents the problem.
I intentionally simplified the issue a lot for you to be able to reproduce it. The change you suggest is not possible in the original code.

**Reply by jamiedemaria** [reply]
can you elaborate on why you can't add the `required_resource_keys={"slack"}` to the op decorator?
